### PR TITLE
feat: implement JsonOpener and CsvOpener

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Due to the Datafusion being highly coupled with `object_store::ObjecStore`. We must find a way to utilize the Datafusion without being coupled with `object_store::ObjecStore`. 
One approach is to implement our own `FileOpener`. Luckily, `CsvOpener` and `JsonOpener` were not complex. And in the future, we also can reuse code from the existing `ParquetReader` to implement `ParquetOpener`.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1041 #1305

### CODE REVIEW

File: src/common/datasource/src/physical_plan/file_format/csv.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+#[derive(Debug, Clone)]
+pub struct CsvOpener {
+    config: Arc<CsvConfig>,
+    object_store: Arc<ObjectStore>,
+    compression_type: CompressionType,
+}
+
+impl CsvOpener {
+    /// Return a new [`CsvOpener`]. The call must ensure [`CsvConfig`].file_schema must correspond to the opening file.
```
Comment (WenyXu) [inline src/common/datasource/src/physical_plan/file_format/csv.rs:74]:
```suggestion
    /// Return a new [`CsvOpener`]. The caller must ensure [`CsvConfig`].file_schema must correspond to the opening file.
```

File: src/common/datasource/src/physical_plan/file_format/file_stream.rs [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                        if let NextOpen::Pending(f) = next_open_future {
+                            if let Poll::Ready(stream) = f.as_mut().poll(cx) {
+                                *next_open_future = NextOpen::Ready(stream);
+                            }
+                        }
+                    }
+                    match ready!(stream.poll_next_unpin(cx)) {
+                        Some(result) => {
+                            // TODO(weny): should we support the projection in FileStream?
+                            // Maybe we can utilize the projection ability of the file opener.
```
Comment (WenyXu) [inline src/common/datasource/src/physical_plan/file_format/file_stream.rs:143]: I guess the answer is Yes. We still need to handle cases like ordering.

**Comment by WenyXu** [comment]
PTAL 😇 @v0y4g3r @waynexia @evenyag

**Comment by v0y4g3r** [comment]
Is doupling with `object_store::ObjectStore` the only barricade that prevents us from using arrow's [`FileOpener`](https://docs.rs/datafusion/22.0.0/datafusion/physical_plan/file_format/trait.FileOpener.html)?

**Comment by WenyXu** [comment]
> Is doupling with `object_store::ObjectStore` the only barricade that prevents us from using arrow's [`FileOpener`](https://docs.rs/datafusion/22.0.0/datafusion/physical_plan/file_format/trait.FileOpener.html)?

Yes 🥲

**Comment by v0y4g3r** [comment]
> > Is doupling with `object_store::ObjectStore` the only barricade that prevents us from using arrow's [`FileOpener`](https://docs.rs/datafusion/22.0.0/datafusion/physical_plan/file_format/trait.FileOpener.html)?
> 
> Yes 🥲

Can we use opendal as the backend of object_store? Looks like there's already [a binding for that](https://github.com/apache/incubator-opendal/issues/1222). 

The reasons why I prefer this approach:
- we don't want to introduce a large patch solely for external table functionality
- the code base is a datafusion fork with object_store replaced and we have to maintain it when upstream updates.

File: src/common/datasource/src/file_format.rs [hunk 3]
```diff
@@ -15,16 +15,95 @@
...
+            loop {
+                if buffered.is_empty() {
+                    if let Some(result) = futures::ready!(upstream.poll_next_unpin(cx)) {
+                        buffered = result?;
+                    };
+                }
+
+                let decoded = decoder.decode(buffered.as_ref())?;
+
+                if decoded == 0 {
```
Comment (v0y4g3r) [inline src/common/datasource/src/file_format.rs:97]: nit: maybe we should add a doc on `ArrowDecoder::decode` method that returning `Ok(0)` means reaching end of file.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Comment by WenyXu** [comment]
waiting for 😅 #1387

**Comment by WenyXu** [comment]
PTAL @waynexia

**Review by waynexia — APPROVED** [review]
LGTM 🚀

**Pull request merged**